src/gc-heap-snapshot.cpp

```diff
@@ -0,0 +1,420 @@
+            }
+            // Escape strings for JSON
+            // TODO
+            print_str_escape_json(stream, str);
+        }
+        jl_printf(stream, "]");
+    }
+};
+
+struct HeapSnapshot {
+public:
+
+// private:
+    vector<Node> nodes;
+    // edges are stored on each from_node
+
+    StringTable names;
+    StringTable node_types;
+    StringTable edge_types;
+    unordered_map<void*, size_t> node_ptr_to_index_map;
+
+    size_t num_edges = 0; // For metadata, updated as you add each edge. Needed because edges owned by nodes.
+};
+
+
+// TODO: Do we need to refer to nodes by their index in the from_node array?
+//size_t find_or_create_node_id(HeapSnapshot& snapshot, string key) {
+//    return find_or_insert_iter(snapshot.nodes_map, key)->second;
+//}
+
+HeapSnapshot *g_snapshot = nullptr;
+
+JL_DLLEXPORT int count_nodes = 0;
+JL_DLLEXPORT int count_edges = 0;
+
+JL_DLLEXPORT void jl_gc_take_heap_snapshot(JL_STREAM *stream) {
+    // Enable snapshotting
+    HeapSnapshot snapshot;
+    g_snapshot = &snapshot;
+    gc_heap_snapshot_enabled = true;
+
+    _add_internal_root(&snapshot);
+
+    // Do GC, which will callback into record_edge_to_gc_snapshot()...
+    jl_gc_collect(JL_GC_FULL);
+
+    // Disable snapshotting
+    gc_heap_snapshot_enabled = false;
+    g_snapshot = nullptr;
+
+    // When we return, the snapshot is full
+    // Dump the snapshot
+    serialize_heap_snapshot(stream, snapshot);
+
+    // Debugging
+    //jl_printf(JL_STDERR, "nodes: %d\n", count_nodes);
+    //jl_printf(JL_STDERR, "edges: %d\n", count_edges);
+}
+
+// adds a node at id 0 which is the "uber root":
+// a synthetic node which points to all the GC roots.
+void _add_internal_root(HeapSnapshot *snapshot) {
+    Node internal_root{
+        snapshot->node_types.find_or_create_string_id("synthetic"),
+        "(internal root)", // name
+        0, // id
+        1, // size
+
+        0, // size_t trace_node_id (unused)
+        0, // int detachedness;  // 0 - unknown,  1 - attached;  2 - detached
+
+        // outgoing edges
+        vector<Edge>(),
+    };
+    snapshot->nodes.push_back(internal_root);
+}
+
+// mimicking https://github.com/nodejs/node/blob/5fd7a72e1c4fbaf37d3723c4c81dce35c149dc84/deps/v8/src/profiler/heap-snapshot-generator.cc#L597-L597
+void record_node_to_gc_snapshot(jl_value_t *a) JL_NOTSAFEPOINT {
+    auto val = g_snapshot->node_ptr_to_index_map.find((void*)a);
+    if (val != g_snapshot->node_ptr_to_index_map.end()) {
+        return;
+        //return &g_snapshot->nodes[val->second];
+    }
+
+    // Insert a new Node
+    size_t self_size = 0;
+    string name = "<missing>";
+    string node_type = "object";
+
+    if (a == (jl_value_t*)jl_malloc_tag) {
+        name = "<malloc>";
+    } else {
+        jl_datatype_t* type = (jl_datatype_t*)jl_typeof(a);
+
+        if ((uintptr_t)type < 4096U) {
+            name = "<unkown>";
+        } else if (type == (jl_datatype_t*)jl_buff_tag) {
+            name = "<buffer>";
```

Comment:
These can only show up in very specific contexts (jl_binding_t and jl_array_t->data), so it might be better to just avoid passing those here and instead charge their size directly to their parent.

Reply:
Yeah, we considered doing that, too, but we figured it makes more sense to report them as "internal" edges, which (we think) tells the javascript frontend that there's _some_ object there, but it's not a normal object and it's opaque. That seems like it's the right fit for this kind of value, too? That way we can still ask questions like "what is the size of all the array buffers," in case that was something we were interested in, i guess?